### PR TITLE
Expressions in meta sections

### PR DESCRIPTION
**Pull request by orodeh**
Here is the the initial code for adding JSON-like objects to meta sections. In sbt project `wdlTransformsDraft3`, there is a new test case for the new constructions. What is the next step?

### CODE REVIEW

**Comment by orodeh** [comment]
@cjllanwarne, I thought I could assign the review to you, but that is not possible with my permissions. Please let me know how to proceed with this PR. 

Thank you!
      Ohad.

**Comment by orodeh** [comment]
The modified hermes grammar, based on (https://github.com/cjllanwarne/wdl/tree/cjl_draft3_grammar_changes), is at:
https://github.com/dnanexus-rnd/wdl/blob/expressions_in_meta_sections/parsers/grammar.hgr

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      a.getAttributeAsVector[MetaValueElement]("values").toValidated.map(MArray)
+
+    case a: Ast if a.getName == "MetaObject" =>
+      (for {
+        mapKvs <- a.getAttributeAsVector[MetaKvPair]("map")
+        asMap = mapKvs.map(kv => kv.key -> kv.value).toMap
+       } yield MObject(asMap)).toValidated
+
+    case other =>
+      Failure(new Exception(s"No conversion defined for Ast $other to MetaValueElement")).toErrorOr
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala]: No Exception instance is needed here - you can just say `s"No conversion defined for Ast $other to MetaValueElement".invalidNel`

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  def convert(astNode: AstNode): ErrorOr[MetaValueElement] = astNode match {
+
+      // This is a primitive type, one of {null, boolean, float, int, string}.
+    case t: Terminal =>
+      (t.getTerminalStr, t.getSourceString) match {
+        case ("integer", i) => Try(MInteger(i.toInt)).toErrorOr
+        case ("float", f) => Try(MFloat(f.toDouble)).toErrorOr
+        case ("boolean", b) => Try(MBoolean(b.toBoolean)).toErrorOr
+        case ("string", s) => MString(s).validNel
+        case ("null", _) => Try(MNull).toErrorOr
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala]: To return a safe valid value, use `MNull.validNel`

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala (hunk 1 above)
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/AstNodeToMetaValueElement.scala]: Also the message should probably be `"No conversion defined for AstNode $other to MetaValueElement"` (`AstNode` is the supertype which could be an `Ast`, `Terminal` or `AstList`)

**Comment by orodeh** [comment]
@cjllanwarne I don't know why the `docker_hash_quay` test is failing, but I don't think this is related to my changes.

**Comment by cjllanwarne** [comment]
@orodeh yeah, I wouldn't worry about that (it sounds like a hash was changed in quay.io for the first time in 2 years and it broke that test. Nothing for us to worry about in this PR!)

I'll wait for the tests to complete again and as long as it's still looking good, I'll merge

**Comment by orodeh** [comment]
That great, thanks!

**Pull request merged**